soaktest/soaktest.go

```diff
@@ -0,0 +1,285 @@
+// ConsumerError combines the metrics name for consumer client-level errors
+const ConsumerError = "consumer.err"
+
+// ProducerError combines the metrics name for producer client-level errors
+const ProducerError = "producer.err"
+
+// FailedToProduceMsg combines the metrics name if failed to produce messages
+const FailedToProduceMsg = "failed.to.produce.message"
+
+var producerDr = "producer.dr"
+var producerDrErr = "producer.dr.err"
+
+const consumerConsumeMSG = "consumer.consume.msg"
+const consumerReceiveError = "consumer.receive.err"
+const latency = "latency"
+const consumerConsumeDupMSG = "consumer.consume.dup.msg"
+const consumerConsumeMissMSG = "consumer.consume.miss.msg"
+const brokerRttP99 = "broker.rtt.p99"
+const brokerRttAvg = "broker.rtt.avg"
+
+// InitLogFiles initials the log folder and file, if the folder or file
+// doesn't exist, create one, otherwise open it directly
+func InitLogFiles(filename string) {
+	path := filepath.Dir(filename)
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		os.Mkdir(path, os.ModePerm)
+	}
+
+	file, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY,
+		0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	InfoLogger = log.New(file, "INFO: ", log.Ldate|log.Ltime|log.Lshortfile)
+	WarningLogger = log.New(file, "WARNING: ",
+		log.Ldate|log.Ltime|log.Lshortfile)
+	ErrorLogger = log.New(file, "ERROR: ", log.Ldate|log.Ltime|log.Lshortfile)
+}
+
+// rttStats extracts broker rtt statistics from the raw map,
+// monitors broker avg and p99 using datadog
+func rttStats(raw map[string]interface{}, topic string) error {
+	for _, broker := range raw["brokers"].(map[string]interface{}) {
+		var parts string
+		var avg float64
+		var p99 float64
+		var nodeid int
+
+		topparsValue := broker.(map[string]interface{})["toppars"]
+		if topparsValue == nil ||
+			len(topparsValue.(map[string]interface{})) == 0 {
+			continue
+		}
+
+		for _, val := range topparsValue.(map[string]interface{}) {
+			if val.(map[string]interface{})["topic"] != topic {
+				continue
+			}
+			partition := val.(map[string]interface{})["partition"]
+			f, ok := partition.(float64)
+			if !ok {
+				err := fmt.Errorf("Failed to convert partition "+
+					"to float64 %v\n", ok)
+				return err
+			}
+			if len(parts) > 0 {
+				parts += ","
+			}
+			parts += fmt.Sprintf("%d", int(f))
+		}
+
+		rttValue := broker.(map[string]interface{})["rtt"]
+		avg, ok := rttValue.(map[string]interface{})["avg"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert avg to float64 %v", ok)
+			return err
+		}
+
+		p99, ok = rttValue.(map[string]interface{})["p99"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert p99 to float64 %v", ok)
+			return err
+		}
+
+		nid, ok := broker.(map[string]interface{})["nodeid"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert nodeid to float64 %v", ok)
+			return err
+		}
+		nodeid = int(nid)
+
+		tags := []string{fmt.Sprintf("broker:%d, type:%s, topic:%s, parts:%s",
```

Comment:
I don't think topic and partitions should  be part of the per-broker rtts since any change in leadership will create a new data series.
What I do in the Python soaks is have one series for the broker rtt, and one series for partition leaders. This allows us to manually map current-partition-leader-broker-rtts in datadog by comparing the two charts.